### PR TITLE
ENH - Implement a Cox scikit-learn-like Estimator

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
follow up of #157, #165, #167 

This wraps the problem of solving a Cox problem in a scikit-learn-like estimator.

### CODE REVIEW

**Comment by Badr-MOUFAD** [comment]
To have scikit-learn compatibility we need to switch convention ``(tm, s) --> y``, as suggested by @PABannier, and opt instead for a ``y`` with two columns. 

With that being said, any idea on how to run ``check_estimator`` on this case?

File: skglm/estimators.py [hunk 1]
```diff
@@ -1159,6 +1162,169 @@ def fit(self, X, y):
...
 
 
+class CoxEstimator(LinearModel):
+    r"""Elastic Cox estimator with Efron and Breslow estimate.
+
+    Refer to :ref:`Mathematics behind Cox datafit <maths_cox_datafit>`
+    for details about the datafit expression. The data convention for the estimator is
+
+    - ``X`` the design matrix with ``n_features`` predictors
+    - ``y`` a two-column vector where the first ``tm`` is of event time occurrences
```
Comment (mathurinm) [inline skglm/estimators.py:1172]: is "a two-column vector" standard?

File: skglm/estimators.py [hunk 2]
```diff
@@ -1159,6 +1162,169 @@ def fit(self, X, y):
...
+      and the second ``s`` is of censoring.
+
+    For L2 Cox, ``l1_ratio=0.``, :ref:`LBFGS <skglm.solvers.LBFGS>`
```
Comment (mathurinm) [inline skglm/estimators.py:1175]:
```suggestion
    For L2-regularized Cox (``l1_ratio=0.``) :ref:`LBFGS <skglm.solvers.LBFGS>`
```

File: skglm/estimators.py [hunk 3]
```diff
@@ -1159,6 +1162,169 @@ def fit(self, X, y):
...
+    is the used solver, otherwise it is :ref:`ProxNewton <skglm.solvers.ProxNewton>`.
+
+    Parameters
+    ----------
+    alpha : float, optional
+        Penalty strength. It must positive ``alpha > 0``.
```
Comment (mathurinm) [inline skglm/estimators.py:1181]:
```suggestion
        Penalty strength. It must strictly positive.
```

File: skglm/estimators.py [hunk 4]
```diff
@@ -1159,6 +1162,169 @@ def fit(self, X, y):
...
+
+    l1_ratio : float, default=0.7
```
Comment (mathurinm) [inline skglm/estimators.py:1183]: Any hint on this default?

File: skglm/estimators.py [hunk 5]
```diff
@@ -1159,6 +1162,169 @@ def fit(self, X, y):
...
+        The ElasticNet mixing parameter, with ``0 <= l1_ratio <= 1``. For
+        ``l1_ratio = 0`` the penalty is an L2 penalty. ``For l1_ratio = 1`` it
+        is an L1 penalty.  For ``0 < l1_ratio < 1``, the penalty is a
+        combination of L1 and L2.
+
+    method : {'efron', 'breslow'}, default='efron'
```
Comment (mathurinm) [inline skglm/estimators.py:1189]: is it the name that lifelines uses?

File: skglm/estimators.py [hunk 6]
```diff
@@ -1159,6 +1162,169 @@ def fit(self, X, y):
...
+        The estimate used for the Cox datafit. Use ``efron`` to
+        handle tied observation.
```
Comment (mathurinm) [inline skglm/estimators.py:1191]:
```suggestion
        handle tied observations.
```

File: skglm/estimators.py [hunk 7]
```diff
@@ -1159,6 +1162,169 @@ def fit(self, X, y):
...
+        X = check_array(
+            X,
+            accept_sparse="csc",
+            order="F",
+            dtype=[np.float64, np.float32],
+            input_name="X",
+        )
+        if y is None:
+            # Needed to pass check estimator. Message error is
+            # copy/past from https://github.com/scikit-learn/scikit-learn/blob/ \
```
Comment (mathurinm) [inline skglm/estimators.py:1259]:
```suggestion
            # copy/paste from https://github.com/scikit-learn/scikit-learn/blob/ \
```

File: skglm/tests/test_estimators.py [hunk 8]
```diff
@@ -260,6 +261,73 @@ def test_CoxEstimator_sparse(use_efron, use_float_32):
...
+    X, y = make_dummy_survival_data(n_samples, n_features, normalize=True,
+                                    with_ties=use_efron, random_state=0)
+
+    estimator_sk = CoxEstimator(
+        alpha, l1_ratio=l1_ratio, method=method, tol=1e-6
+    ).fit(X, y)
+    w_sk = estimator_sk.coef_
+
+    # fit lifeline estimator
+    stacked_tm_s_X = np.hstack((y, X))
```
Comment (mathurinm) [inline skglm/tests/test_estimators.py:288]: YMMV: no need to define variable if used only once and code is straightforward

File: skglm/tests/test_estimators.py [hunk 9]
```diff
@@ -260,6 +261,73 @@ def test_CoxEstimator_sparse(use_efron, use_float_32):
...
+
+    X, y = make_dummy_survival_data(n_samples, n_features, X_density=0.1,
+                                    with_ties=use_efron, random_state=0)
+
+    estimator_sk = CoxEstimator(
+        alpha, l1_ratio=l1_ratio, method=method, tol=1e-9
+    ).fit(X, y)
+    stop_crit = estimator_sk.stop_crit_
+
+    np.testing.assert_allclose(stop_crit, 0., atol=1e-8)
```
Comment (mathurinm) [inline skglm/tests/test_estimators.py:324]: in all rigor we'd like to use assert_array_less to check that it's lower than 1e-9

File: skglm/tests/test_lbfgs_solver.py [hunk 10]
```diff
@@ -26,7 +26,7 @@ def test_lbfgs_L2_logreg(X_sparse):
     # fit L-BFGS
     datafit = compiled_clone(Logistic())
     penalty = compiled_clone(L2(reg))
-    w, *_ = LBFGS(tol=1e-12).solve(X, y, datafit, penalty)
+    w, *_ = LBFGS(tol=1e-5).solve(X, y, datafit, penalty)
```
Comment (mathurinm) [inline skglm/tests/test_lbfgs_solver.py:29]: is this a leftover that went through accidentally or is there a reason for the change?

File: skglm/estimators.py (hunk 3 above)
Comment (Badr-MOUFAD) [reply skglm/estimators.py:1181]: As far as I know, _positive_ means being ``> 0``, whereas _non-negative_ means ``>= 0``.

cf. https://math.stackexchange.com/questions/18464/is-positive-the-same-as-non-negative

File: skglm/estimators.py (hunk 4 above)
Comment (Badr-MOUFAD) [reply skglm/estimators.py:1183]: It should be `0.5` to align with the existing codebase and `scikit-learn`.

File: skglm/estimators.py (hunk 5 above)
Comment (Badr-MOUFAD) [reply skglm/estimators.py:1189]: `lifelines` will automatically fall to using `efron` when data contains tied observations (the user cannot contest)

[Here, baseline_estimation_method,](https://lifelines.readthedocs.io/en/latest/fitters/regression/CoxPHFitter.html?highlight=coxph#lifelines.fitters.coxph_fitter.CoxPHFitter)  is the occurrence of the name ``breslow``. On the other hand, ``efron`` is [hard-coded in the codebase](https://github.com/CamDavidsonPilon/lifelines/blob/bcf040caa5a8bc706de316392abf9d4c3626a4d4/lifelines/fitters/coxph_fitter.py#L1380)

File: skglm/estimators.py (hunk 3 above)
Comment (mathurinm) [reply skglm/estimators.py:1181]: yeah it's a shit way of naming things so let's go the french way

File: skglm/estimators.py (hunk 1 above)
Comment (mathurinm) [reply skglm/estimators.py:1172]: what's a "two-column vector" ?

File: skglm/estimators.py [hunk 11]
```diff
@@ -1159,6 +1162,169 @@ def fit(self, X, y):
...
+    - ``y`` a two-column vector where the first ``tm`` is of event time occurrences
+      and the second ``s`` is of censoring.
+
+    For L2-regularized Cox (``l1_ratio=0.``) :ref:`LBFGS <skglm.solvers.LBFGS>`
+    is the used solver, otherwise it is :ref:`ProxNewton <skglm.solvers.ProxNewton>`.
+
+    Parameters
+    ----------
+    alpha : float, optional
+        Penalty strength. It must be strictly positive ``alpha > 0``.
```
Comment (mathurinm) [inline skglm/estimators.py:1181]:
```suggestion
        Penalty strength. It must be strictly positive.
```

File: skglm/estimators.py (hunk 1 above)
Comment (Badr-MOUFAD) [reply skglm/estimators.py:1172]: It should be array instead, thanks!

**Pull request merged**